Download and install newer versions of bossac during installation

FHeilmann opened the pull request:
The `bossac` version included in the debian and ubuntu repositories is very old. One can't, for example, use it with the sam4e8e MCU. This change pulls a more recent version (1.8) from the git repository, compiles only the CLI executable and installs it in `/usr/bin`. I avoided the most recent version (1.9) for now because there are reports of that version bricking some devices by overwriting the bootloader unless the user passes a specific offset to the utility.

Let me know if you want me to change anything. I can also provide the installation as a separate script.

KevinOConnor commented:
Thanks.  A few days back, I compiled bossac-1.8 as part of my exploration into the samd21 chips.  Unfortunately, when I tried to flash my Arduino Due vis bossac-1.8 I got a failure with a cryptic error message.  When I reverted back to bossac-1.6.1 the problem went away.  I'll recheck (it may just require a change to the command line flags), but it didn't look encouraging.

Separately, I don't think the install script should put files in /usr/bin, as that goes against the guidelines of many OS package systems.  If we need to build 3rd party programs I think we should limit them to the user's directory or to the klipper directory itself.  Also, be aware that the install-octopi.sh script may be rerun (for example, on a software upgrade), so we don't want to download and build packages that are already present.

-Kevin

FHeilmann commented:
What a mess, the SAM4e does not work with <1.8 so we'll effectively have to keep two versions around. How do you feel about creating a /tools folder in the klipper folder and storing bossac over there?

FHeilmann commented:
@KevinOConnor I changed the script according to your second request. The script now installs everything into `$KLIPPER_DIR/tools/bossa` instead of `/usr/bin`. The script also checks whether this directory exists so the building process isn't rerun if the install script is rerun.

I added the latest version from the arduino branch of the `schumatech/BOSSA` repository. Can you check whether this version (installed to `tools/bossa/bossac1.7-arduino`) can flash the Arduino Due?

FHeilmann commented:
I just did some testing on a borrowed Due board.

Via the **programming** port, using 

```
stty -F /dev/ttyACM0 1200
tools/bossa/bossac1.7-arduino -p /dev/ttyACM0 -w -v -e out/klipper.bin 
```

successfully flashes the board. bossac1.8 also works using the following command

tools/bossa/bossac1.8 -p /dev/ttyACM0 -w -v -e out/klipper.bin

but only via the **native** port.

KevinOConnor commented:
Thanks.  I wonder if it would be simpler to include the bossac v1.8 code in the lib/ directory.  I put up a work-bossac-20180816 branch (untested) to demonstrate the idea.  Let me know what you think.

-Kevin

FHeilmann commented:
Sure. This works too if you prefer it this way. Only thing I can think of is to double check that bossac doesn't have any additional dependencies that the install script should handle.

KevinOConnor commented:
Okay - I got a chance to play a little with this.  I was able to get the latest bossac (v1.9) to reliably flash my Arduino Due.  For some reason, the cpu reset (-R) can't be run from the main flashing sequence on the Due, but it works fine if it's run separately.

I also tested this on my raspberry pi - bossac seems to compile with the current dependencies.

I've updated the work-bossac-20180816 accordingly.  Could you verify that everything still works on your side with that branch?  If everything looks okay, I think we can merge this.

-Kevin

FHeilmann commented:
I'll test it. You should also be able to use that bossac to flash the samd21


If you're doing so, be extra careful because bossac may overwrite the bootloader (see here https://github.com/shumatech/BOSSA/issues/79)

KevinOConnor commented:
Thanks.

FYI, I have an Arduino M0 clone that came with a bootloader that uses the stk500v2 protocol.  As near as I can tell, the SAMD21 needs a bootloader to flash (there is no builtin rom like on the SAM3/SAM4).  As near as I can tell, the only way to flash the SAMD21 without a bootloader is to use a jtag like hardware device.

It seems the Arduino Zero has a bootloader that supports bossac, and it seems the Zero also has a separate programming chip (with its own usb port) that can flash the SAMD21 via the bossac protocol (it appears to translate the flash requests to a jtag port on the main chip).

FWIW, I did find an article explaining how one could flash a bootloader to the SAMD21 using an rpi as the hardware programmer: https://learn.adafruit.com/programming-microcontrollers-using-openocd-on-raspberry-pi/wiring-and-test

-Kevin

FHeilmann commented:
Basically my reasoning is that 1.9 can overwrite the bootloader if you don't pass the right arguments while 1.8 has the offset hardcoded so you can can't screw up your board no matter what you do. I'll close this PR and am looking forward to this going master. Thanks for your efforts 😊

KevinOConnor commented:
Thanks.  I committed work-bossac-20180816 to the master branch.

Okay, thanks for the info.  The one nice thing about v1.9 is that it fixes a bunch of compiler warnings.  It also adds the "-a" flag which is useful on the Due.

-Kevin
